### ISSUE TITLE
Clickable keyword searching [Feature] 

### ISSUE THREAD

**Issue by break6700** [Zettlr/Zettlr#2443]
### Checklist

- [X] Zettlr does not possess this feature (if it does, but the feature doesn't work out for you, propose an enhancement instead.)
- [X] Zettlr's current features cannot be used to approximate this behaviour (otherwise, propose an enhancement instead.)
- [X] I have searched the documentation to make sure I did not overlook it.
- [X] I have searched both open and closed issues to see if someone has already proposed this.

### Feature Proposal

Cmd/Ctrl-clicking YAML frontmatter keywords performs a search for that word.

### Example Workflow

Cmd/Ctrl-clicking YAML frontmatter keywords performs a search for that word (similar to a '#' tag).

### Caveats

Can't think of any.

### Do you wish to attempt implementing this yourself?

No

### Platforms

- [X] Windows
- [X] macOS
- [X] Linux

### Additional Information

This is following a conversation on Discord.

**Comment by nathanlesage** [comment]
I'm not yet putting this into the 2.0 project, since I don't know whether or not I can make it until then; feel free to keep an eye on this issue and ping me before 2.0 if needed!

**Comment by alpcantal** [comment]
It doesn't have to be strictly in YAML so long as there is "clickable search". 

An improved version of the "search internal links" feature that searches through the file filter may be a better option. Not a programmer, but I'd think that would be easier to implement.

**Comment by nathanlesage** [comment]
> It doesn't have to be strictly in YAML so long as there is "clickable search".

This issue here simply refers to the fact that keywords in YAML frontmatter are not yet clickable. It only wants to put the YAML frontmatter keywords on par with in-text tags, which are already clickable. Your proposal of filtering is unaffected by this one!

**Comment by Julianoe** [comment]
@nathanlesage tags and keywords are marked as "clickable" as of this issue https://github.com/Zettlr/Zettlr/issues/607 and this commit https://github.com/Zettlr/Zettlr/commit/9d633836f57f9b2ab7ad9cc298912b2190aa4904
But in Zettlr 4.1.0 they are not clickable anymore.
Is it normal? Should I open another issue or is this issue dedicated to following this problem?

Summary: tags and keywords in YAML are not (no longer) clickable.

**Comment by nathanlesage** [comment]
That has been an issue for quite some time, and there is an issue open somewhere on this tracker for that.

**Comment by BabakFarshchian** [comment]
See also #4498 . Duplicate?

**Comment by nathanlesage** [comment]
Appears like it. It's time we get to it ... I'm happy for PRs!

**Comment by nathanlesage** [comment]
I think I have found a solution, but it seems a bit silly so @benniekiss I would like to ask you if you have any opinion on that approach, and maybe know of a nicer way to solve this.

The main issue we currently have is that the parser does not give us access to the actual YAML properties in the frontmatter via the syntax tree. This means that we can't really figure out what it is that someone has clicked in the frontmatter.

The way I would solve this now goes as following:

1. In the frontmatter block parser, change the emitted syntax nodes to include a dummy:
   ```ts
   const wrapperNode = ctx.elt('YAMLFrontmatter', 0, ctx.lineStart + 3, [
      ctx.elt('YAMLFrontmatterStart', 0, 3),
      // New dummy node as child of CodeText
      ctx.elt('CodeText', 4, ctx.lineStart - 1 [ctx.elt('DummyNode', 4, ctx.lineStart - 1)]),
      ctx.elt('YAMLFrontmatterEnd', ctx.lineStart, ctx.lineStart + 3)
    ])
   ```
2. Then, adjust the wrap parser to not use an overlay mode, and instead directly mount the tree in lieu of said dummy node:
   ```ts
   if (node.type.name === 'DummyNode' && node.node.parent?.name === 'CodeText' && node.node.parent?.parent?.name === 'YAMLFrontmatter') {
      return { parser }
    } else {
      return null
    }
   ```

By doing it like this we can achieve three things:

1. Do not mess with any code that expects a `CodeText` node as a child of the `YAMLFrontmatter` node (such as the code background renderer -- while testing this I accidentally nuked the ability of the app to parse frontmatters)
2. The nodes emitted by the yaml parser will actually be present in the syntaxTree and are as such available to, in this case, the `clickHandler` that can check whether the user has clicked on a tag or keyword in the frontmatter
3. Do all of this in the least invasive way possible, with changes only to [frontmatter-parser.ts](source/common/modules/markdown-editor/parser/frontmatter-parser.ts), the node definition file, and the click handler.

But again, this seems a bit silly, so the question is: is there a nicer way? If not, I'd do it so.

**Comment by benniekiss** [comment]
Hm. I'll think about this, but I think it may be worth migrating to the `lang-yaml` frontmatter parser instead of the custom one we have. I'm pretty sure that would achieve the same with far less code:

https://code.haverbeke.berlin/codemirror/lang-yaml#yamlfrontmatter

**Comment by nathanlesage** [comment]
Huh? Are you sure we aren't already using that one?

**Comment by nathanlesage** [comment]
Yes, just checked: the wrapper already uses the official YAML parser. The problem is rather that we don't have the syntax tree available of that parser.

I mean, another option would be to remove the frontmatter parser entirely from the markdown parser's wrapper and do a manual compute-and-mount directly in the block parser, so that the CodeText just has that tree directly as children. That might be more efficient.

**Comment by benniekiss** [comment]
I'll look into it more, but I believe the reason yaml nodes are not showing up in the syntax tree currently is exactly because of all the wrapping we are doing. The codemirror [`yamlfrontmatter`](https://code.haverbeke.berlin/codemirror/lang-yaml#yamlfrontmatter) is a top-level wrapper that would wrap the markdown parser, and I believe this would expose the yaml nodes.

Briefly looking over the current implementation, I think it could probably be simplified a lot, removing the nested `CodeText` and further nested `Yaml` nodes. I think the yaml parser can just be used to parse the frontmatter and have the nodes pushed directly into the tree.

EDIT: I realized my last point is what you were saying about remounting the nodes. I think that would probably be best.

**Comment by nathanlesage** [comment]
Ah True, I forgot to mention that I don't want the YAML frontmatter wrapper, because I want to wrap the entire parser into my criticmarkup parser once I get to it.

**Comment by nathanlesage** [comment]
Bonjour! I think I have finally managed to do it. I have done quite a lot of damage to the various parsers in the meantime, so I will open a PR with the changes soon, and wait for a review. There are still a few oddities in there, but fundamentally, it works.
